### PR TITLE
OpenRouter based LLM models

### PR DESCRIPTION
**Pull request by bauersimon**
- base structure for LLM models
- [OpenRouter](https://openrouter.ai) as cloud provider

### CODE REVIEW

File: model/cloud/openrouter.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package cloud
```
Comment (zimmski) [inline model/cloud/openrouter.go:1]: cloud->provider

becaus this package will holds our providers.

Also, please move on the same level as "model", so we can do subpackages for models.

File: model/cloud/mock.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package cloud
+
+import "github.com/stretchr/testify/mock"
+
+// MockProvider is a mocked Provider.
+type MockProvider struct {
```
Comment (zimmski) [inline model/cloud/mock.go:6]: Needs to be moved into a subpackage "testing"

File: model/cloud/openrouter.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/sashabaranov/go-openai"
+)
+
+type openRouterProvider struct {
+	// client is the client to query the LLM model.
+	client *openai.Client
+}
+
+// OpenRouterToken is the "OPENROUTER_TOKEN" environment variable.
+var OpenRouterToken = os.Getenv("OPENROUTER_TOKEN")
```
Comment (zimmski) [inline model/cloud/openrouter.go:17]: Needs to be a parameter. The Env variable should be resolved on the "cmd" level not in the model/router/... code.

File: model/cloud/openrouter.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	config.BaseURL = "https://openrouter.ai/api/v1"
+	client := openai.NewClientWithConfig(config)
+
+	return &openRouterProvider{
+		client: client,
+	}
+}
+
+// Query queries the LLM.
+func (o *openRouterProvider) Query(request string, model string) (response string, err error) {
```
Comment (zimmski) [inline model/cloud/openrouter.go:31]: model is a context parameter, so please put it before the request.

Comment (zimmski) [inline model/cloud/openrouter.go:31]: reqeuest -> promptText

File: model/cloud/openrouter.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	apiResponse, err := o.client.CreateChatCompletion(
+		context.Background(),
```
Comment (zimmski) [inline model/cloud/openrouter.go:33]: Please make the "context" a parameter (first parameter) so we can do the timeouts from outside right away.

File: model/cloud/openrouter.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+			Messages: []openai.ChatCompletionMessage{
+				{
+					Role:    openai.ChatMessageRoleUser,
+					Content: request,
+				},
+			},
+		},
+	)
+	if err != nil {
+		return "", err
```
Comment (zimmski) [inline model/cloud/openrouter.go:45]: Add traces to all errors (see the symflower model package for an example)

File: model/cloud/provider.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package cloud
+
+// Provider is a cloud provider to query LLMs.
```
Comment (zimmski) [inline model/cloud/provider.go:3]: -> `Provider defines a provider to query models such as LLMs.` ? to get rid of the word cloud

File: model/cloud/provider.go [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+type Provider interface {
+	// Query queries the LLM with the given model name.
```
Comment (zimmski) [inline model/cloud/provider.go:5]: -> `Query queries the provider with the given model name.`

File: model/llm.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/symflower/eval-symflower-codegen-testing/model/cloud"
+	"github.com/symflower/eval-symflower-codegen-testing/util"
+	"github.com/zimmski/osutil/bytesutil"
```
Comment (zimmski) [inline model/llm.go:11]: Divide between imports

stdlib

external

internal

File: model/llm.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+// llm represents a LLM model accessed via a cloud provider.
+type llm struct {
+	// provider is the client to query the LLM model.
+	provider cloud.Provider
+	// model holds the identifier string for the LLM model.
+	model string
+}
+
+var modelIdentifiers = []string{
```
Comment (zimmski) [inline model/llm.go:22]: Registring of the openrouter models should happen in the file of openrouter.

File: model/llm.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+// NewLLMModel returns an LLM model corresponding to the given ID which is queried via the given provider.
+func NewLLMModel(provider cloud.Provider, id string) Model {
+	return &llm{
+		provider: provider,
+		model:    id,
+	}
+}
+
+type contextGenerateTestPrompt struct {
```
Comment (zimmski) [inline model/llm.go:41]: ->llmGenerateTestForFilePromptData

and pleas document everything.

File: util/parse.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/zimmski/osutil/bytesutil"
+)
+
+var (
+	codeTagRe           = regexp.MustCompile("(^|\n)\\s*```\\w*($|\n)")
+	codeTagDuplicatedRe = regexp.MustCompile("```(\\s|\n)*```")
+)
+
+// ParseModelResponse parses code from a model's response.
+func ParseModelResponse(response string) (code string) {
```
Comment (zimmski) [inline util/parse.go:16]: Please move this to a model/prompt package

->ParseResponse

File: model/cloud/openrouter.go (hunk 1 above)
Comment (bauersimon) [reply model/cloud/openrouter.go:1]: With this change we have `provider.Provider`. Should I rename `provider.Interface`?

Comment (zimmski) [reply model/cloud/openrouter.go:1]: provider.Provider is fine. We have it in language and model too. Internally we do that too.

File: cmd/eval-symflower-codegen-testing/cmd/evaluate.go [hunk 13]
```diff
@@ -24,6 +25,8 @@ type Evaluate struct {
 	Models []string `long:"model" description:"Evaluate with this model. By default all models are used."`
 	// TestdataPath determines the testdata path where all repositories reside grouped by languages.
 	TestdataPath string `long:"testdata" description:"Path to the testdata directory where all repositories reside grouped by languages." default:"testdata/"`
+
+	ProviderTokens map[string]string `long:"tokens" description:"API tokens for model providers (of the form '$provider:$token,...')." env:"PROVIDER_TOKEN"`
```
Comment (zimmski) [inline cmd/eval-symflower-codegen-testing/cmd/evaluate.go:29]: Always document

File: model/llm/llm.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+// llmGenerateTestForFilePromptData is the template for generating an LLM test generation prompt.
+var llmGenerateTestForFilePromptData = template.Must(template.New("templateGenerateTestPrompt").Parse(bytesutil.StringTrimIndentations(`
+	{{ .Prompt }}
+	` + "```" + `
+	{{ .Code }}
+	` + "```" + `
+`)))
+
+// GenerateTestsForFile implements Model.
```
Comment (zimmski) [inline model/llm/llm.go:54]: Always copy the docstring of the interface and keep it in sync. we do this internally too.

File: model/llm/llm.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+	// TODO Ask the model for the test file name or compute it in a more sophisticated manner.
+	fileExtension := filepath.Ext(filePath)
+	testFilePath := filepath.Join(repositoryPath, strings.TrimSuffix(filePath, fileExtension)+"_test"+fileExtension)
+
+	return os.WriteFile(testFilePath, []byte(testContent), 0644)
+}
+
+// ID implements Model.
+func (m *llm) ID() (id string) {
```
Comment (zimmski) [inline model/llm/llm.go:85]: Keep order of methods in sync with the interface an do an explicit interface declaration

File: model/llm/llm.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+	` + "```" + `
+`)))
+
+// GenerateTestsForFile implements Model.
+func (m *llm) GenerateTestsForFile(repositoryPath string, filePath string) (err error) {
+	data, err := os.ReadFile(filepath.Join(repositoryPath, filePath))
+	if err != nil {
+		return err
+	}
+	fileContent := strings.TrimSpace(string(data)) // Remove trailing and leading spaces.
```
Comment (zimmski) [inline model/llm/llm.go:60]: Remove obvious comment.

**Pull request merged**